### PR TITLE
[MRG] Separate entity checks from suffix to allow BIDSpath to represent dataset_description.json file

### PR DESCRIPTION
**Pull request by adam2392**
PR Description
--------------

Closes: #727 

I think those checks still should be applied to BIDs entities, but the suffix is allowed to be "overwritten".

It should be noted that a consequence is the ability to represent arbitrary suffixes again if one really is "hacking". E.g. `suffix=desc-derivative_ieeg, check=False` will work, but isn't obviously a documented or intended usage rn.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

**Comment by agramfort** [comment]
CIs complain @adam2392 

can you see what's going on? thx

**Comment by adam2392** [comment]
https://4228-89170358-gh.circle-artifacts.com/0/dev/generated/mne_bids.BIDSPath.html#mne_bids.BIDSPath

**Review by agramfort — APPROVED** [review]
I'll let @hoechenberger merge if he's happy

File: mne_bids/path.py [hunk 1]
```diff
@@ -571,9 +583,16 @@ def update(self, *, check=None, **kwargs):
...
+            # XXX: suffix can be overridden with non-BIDS characters
+            # which allows `dataset_description.json` to be represented
+            if key != 'suffix':
+                # check if there are any characters not allowed
+                if val is not None and key != 'root':
+                    _check_key_val(key, val)
+            elif self.check:
+                # check if there are any characters not allowed
+                if val is not None and key != 'root':
+                    _check_key_val(key, val)
```
Comment (sappelhoff) [inline mne_bids/path.py:595]: Isn't this equivalent?

```suggestion
            if key == 'suffix' and not self.check:
                continue
            if val is not None and key != 'root':
                _check_key_val(key, val)
```

Looks much simpler to me

**Review by sappelhoff — CHANGES_REQUESTED** [review]

File: mne_bids/path.py [hunk 2]
```diff
@@ -262,6 +262,18 @@ class BIDSPath(object):
...
+    -----
+    BIDS entities are separated generally with a ``"_"`` character, while
+    entity key/value pairs are separated with a ``"-"`` character (e.g.
+    ``subject``, ``session``, ``task``, etc.). There are checks performed
+    to make sure that there are no ``'-'``, ``'_'``, or ``'/'`` characters
+    associated with any of these entities.
+
+    To represent a file such as the ``dataset_description.json`` file,
+    one can set ``check=False``, and pass in ``suffix='dataset_description'``
+    and ``extension='.json'`` to the BIDSPath.
```
Comment (hoechenberger) [inline mne_bids/path.py:276]:
```suggestion
    To represent a filename such as ``dataset_description.json``,
    one can set ``check=False``, and pass ``suffix='dataset_description'``
    and ``extension='.json'``.
```

File: doc/whats_new.rst [hunk 3]
```diff
@@ -43,6 +43,7 @@ Enhancements
 - More detailed error messages when trying to write modified data via :func:`mne_bids.write_raw_bids`, by `Richard Höchenberger`_ (:gh:`719`)
 - If ``check=True``, :class:`mne_bids.BIDSPath` now checks the ``space`` entity to be valid according to BIDS specification Appendix VIII, by `Stefan Appelhoff`_ (:gh:`724`)
 - ``BIDSPath.root`` now automatically expands ``~`` to the user's home directory, by `Richard Höchenberger`_ (:gh:`725`)
+- ``dataset_description.json`` file is now able to be represented as a ``BIDSPath``, by passing in ``check=False`` and ``suffix='dataset_description'``, by `Adam Li`_ (:gh:`729`)
```
Comment (hoechenberger) [inline doc/whats_new.rst:46]:
```suggestion
- Arbitrary file names can now be represented as a `BIDSPath`` by passing the entire name as ``suffix`` and setting ``check=False``, by `Adam Li`_ (:gh:`729`)
```

**Review by hoechenberger — APPROVED** [review]
LGTM once we're green

**Review by sappelhoff — APPROVED** [review]

**Comment by agramfort** [comment]
@adam2392 you have real CI failures apparently

File: mne_bids/path.py [hunk 4]
```diff
@@ -571,6 +583,10 @@ def update(self, *, check=None, **kwargs):
 
         # error check entities
         for key, val in kwargs.items():
+            # suffix may skip a check if check=False to allow things like
+            # dataset_description.json
+            if key == 'suffix' and not self.check:
+                continue
```
Comment (sappelhoff) [inline mne_bids/path.py:589]:
```suggestion
                pass
                
```

**Comment by adam2392** [comment]
Shall I merge this then? (woot looks like I have squash and merge ability now :p)

**Comment by hoechenberger** [comment]
Thanks, @adam2392!

**Pull request merged**